src/HealthChecks.UI.K8s.Operator/Handlers/DeploymentHandler.cs

```diff
@@ -0,0 +1,127 @@
+using k8s;
+using k8s.Models;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace HealthChecks.UI.K8s.Operator.Handlers
+{
+    public class DeploymentHandler
+    {
+        private readonly IKubernetes _client;
+        private readonly ILogger<K8sOperator> _logger;
+
+        public DeploymentHandler(IKubernetes client, ILogger<K8sOperator> logger)
+        {
+            _client = client ?? throw new ArgumentNullException(nameof(client));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public async Task<V1Deployment> Get(HealthCheckResource resource)
+        {
+            var deploymentList = await _client.ListNamespacedDeploymentAsync(resource.Metadata.NamespaceProperty,
+                labelSelector: $"resourceId={resource.Metadata.Uid}");
+            return deploymentList.Items.FirstOrDefault();
+        }
+
+        public async Task<V1Deployment> GetOrCreateAsync(HealthCheckResource resource)
+        {
+            var deployment = await Get(resource);
+            if (deployment != null) return deployment;
+
+            try
+            {
+                var deploymentResource = Build(resource);
+                var response =
+                    await _client.CreateNamespacedDeploymentWithHttpMessagesAsync(deploymentResource,
+                        resource.Metadata.NamespaceProperty);
+                deployment = response.Body;
+
+                _logger.LogInformation("Deployment {deployment} has been created", deployment.Metadata.Name);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError("Error creating deployment: {error}", ex.Message);
+            }
+
+            return deployment;
+
+        }
+
+        public async Task Delete(HealthCheckResource resource)
+        {
+            try
+            {
+                await _client.DeleteNamespacedDeploymentAsync($"{resource.Spec.Name}-deploy",
+                    resource.Metadata.NamespaceProperty);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError("Error deleting deployment for hc resource: {name} - err: {error}", resource.Spec.Name, ex.Message);
+            }
+        }
+
+        public V1Deployment Build(HealthCheckResource resource)
+        {
+
+            var metadata = new V1ObjectMeta
+            {
+                Labels = new Dictionary<string, string>
```

Comment:
It would be better to set `V1ObjectMeta.OwnerReferences` to point to the `HealthCheckResource` instead of relying on labels to find the deployments and services.

Reply:
Yes, definitely better idea, let me change it

Reply:
Do you mean like this, right?

```csharp
            var metadata = new V1ObjectMeta
            {
                OwnerReferences = new List<V1OwnerReference>
                {
                    new V1OwnerReference { Uid = resource.Metadata.Uid, ApiVersion = resource.ApiVersion, Kind = resource.Kind }
                },
                Labels = new Dictionary<string, string>
                {
                    ["app"] = resource.Spec.Name
                },
                Name = $"{resource.Spec.Name}-deploy",
                NamespaceProperty = resource.Metadata.NamespaceProperty
            };
```

Reply:
I do not see any parameter to filter by owner reference UID's in _client.ListNamespacedDeploymentAsync, so this means we need to get all services in namespace and filter